Source Generator Error when Overriding Inherited Tests

## Description

When overriding/skipping an inherited test method in a derived class that uses `[InheritsTests]`, the source generator fails with an `ArgumentException` due to duplicate hint names.

## Error Message

```
CSC : warning CS8785: Generator 'TestMetadataGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'ArgumentException' with message 'The hintName 'TUnitSkipIssue_Tests_ImplementationATests_AdvancedFeature.g.cs' of the added source file must be unique within a generator. (Parameter 'hintName')'.
```

## Use Case

Multiple implementations of the same interface/abstraction need to be tested. To avoid duplicating test code, an abstract base test class contains all test cases. Each implementation has a derived test class with `[InheritsTests]` to run the tests against that specific implementation.

Some implementations don't support all features, so certain tests need to be skipped in specific derived classes. Overriding the test method with `[Skip]` seems like a natural approach, but the build fails even when just overriding without `[Skip]`.

## Minimal Reproduction

```csharp
public abstract class BaseServiceTests
{
    [Test]
    public async Task BasicFeature()
    {
        await Task.CompletedTask;
    }

    [Test]
    public async Task AdvancedFeature()
    {
        await Task.CompletedTask;
    }
}

[InheritsTests]
public class ImplementationATests : BaseServiceTests
{
    [Test]
    [Skip("Implementation A does not support advanced feature")]
    public new Task AdvancedFeature()
    {
        return Task.CompletedTask;
    }
}
```

*(Note: The error occurs with or without the `[Skip]` attribute)*

## Expected Behavior

The source generator should recognize method hiding with the `new` keyword and only register the derived version for execution. This would allow either replacing the test implementation or skipping it with `[Skip]` (preferred).

## Actual Behavior

The build fails because the generator attempts to create two files with the same hint name - one for the inherited test and one for the overridden test.

## Environment

- TUnit Version: 1.0.78 (issue also occurs in 0.80.00)
- .NET Version: 10.0

## Regression Information

This behavior worked correctly in **version 0.79.0**. The issue was introduced in **version 0.80.0** and persists in version 1.0.78.

## Question

Is using `[InheritsTests]` with method hiding the intended approach for skipping tests in specific implementations? If not, what is the recommended pattern for this scenario?

--- a/TUnit.Core.SourceGenerator/Generators/TestMetadataGenerator.cs
+++ b/TUnit.Core.SourceGenerator/Generators/TestMetadataGenerator.cs
@@ -2341,11 +2341,83 @@ public sealed class TestMetadataGenerator : IIncrementalGenerator
     }
 
 
+    private static bool IsMethodHiding(IMethodSymbol derivedMethod, IMethodSymbol baseMethod)
+    {
+        // Must have same name
+        if (derivedMethod.Name != baseMethod.Name)
+        {
+            return false;
+        }
+
+        // Must NOT be an override (overrides are different from hiding)
+        if (derivedMethod.IsOverride)
+        {
+            return false;
+        }
+
+        // Must have matching parameters
+        if (!ParametersMatch(derivedMethod.Parameters, baseMethod.Parameters))
+        {
+            return false;
+        }
+
+        // Derived method's containing type must be derived from base method's containing type
+        var derivedType = derivedMethod.ContainingType;
+        var baseType = baseMethod.ContainingType;
+
+        // Can't hide yourself
+        if (SymbolEqualityComparer.Default.Equals(derivedType.OriginalDefinition, baseType.OriginalDefinition))
+        {
+            return false;
+        }
+
+        // Check if derived type inherits from base type
+        var current = derivedType.BaseType;
+        while (current is not null && current.SpecialType != SpecialType.System_Object)
+        {
+            if (SymbolEqualityComparer.Default.Equals(current.OriginalDefinition, baseType.OriginalDefinition))
+            {
+                return true;
+            }
+            current = current.BaseType;
+        }
+
+        return false;
+    }
+
     private static List<IMethodSymbol> CollectInheritedTestMethods(INamedTypeSymbol derivedClass)
     {
-        return derivedClass.GetMembersIncludingBase().OfType<IMethodSymbol>()
+        var allTestMethods = derivedClass.GetMembersIncludingBase()
+            .OfType<IMethodSymbol>()
             .Where(m => m.GetAttributes().Any(attr => attr.IsTestAttribute()))
             .ToList();
+
+        // Find methods declared directly on the derived class
+        var derivedClassMethods = allTestMethods
+            .Where(m => SymbolEqualityComparer.Default.Equals(m.ContainingType.OriginalDefinition, derivedClass.OriginalDefinition))
+            .ToList();
+
+        // Filter out base methods that are hidden by derived class methods or declared directly on derived class
+        var result = new List<IMethodSymbol>();
+        foreach (var method in allTestMethods)
+        {
+            // Skip methods declared directly on derived class
+            // (they're handled by regular test registration)
+            if (SymbolEqualityComparer.Default.Equals(method.ContainingType.OriginalDefinition, derivedClass.OriginalDefinition))
+            {
+                continue;
+            }
+
+            // Check if this base method is hidden by any derived class method
+            var isHidden = derivedClassMethods.Any(derived => IsMethodHiding(derived, method));
+
+            if (!isHidden)
+            {
+                result.Add(method);
+            }
+        }
+
+        return result;
     }
 
     private static IMethodSymbol? FindConcreteMethodImplementation(INamedTypeSymbol derivedClass, IMethodSymbol baseMethod)

--- a/TUnit.TestProject/Bugs/3813/BaseServiceTests.cs
+++ b/TUnit.TestProject/Bugs/3813/BaseServiceTests.cs
@@ -1,0 +1,16 @@
+namespace TUnit.TestProject.Bugs._3813;
+
+public abstract class BaseServiceTests
+{
+    [Test]
+    public async Task BasicFeature()
+    {
+        await Task.CompletedTask;
+    }
+
+    [Test]
+    public async Task AdvancedFeature()
+    {
+        await Task.CompletedTask;
+    }
+}

--- a/TUnit.TestProject/Bugs/3813/ImplementationATests.cs
+++ b/TUnit.TestProject/Bugs/3813/ImplementationATests.cs
@@ -1,0 +1,12 @@
+namespace TUnit.TestProject.Bugs._3813;
+
+[InheritsTests]
+public class ImplementationATests : BaseServiceTests
+{
+    [Test]
+    [Skip("Implementation A does not support advanced feature")]
+    public new Task AdvancedFeature()
+    {
+        return Task.CompletedTask;
+    }
+}

